src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -357,15 +359,40 @@ void SetArrayItemNode(object d)
                         }
                         else if (node is ValueExecutionNode valueNode)
                         {
-                            node.Result = valueNode.GraphType.Serialize(d)
-                                ?? throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                            bool defaultException = false;
+                            try
+                            {
+                                object result = valueNode.GraphType.Serialize(d);
+                                if (result == null)
+                                {
+                                    defaultException = true;
+                                    if (listType.ResolvedType is NonNullGraphType)
+                                        throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                                    else
+                                        throw new ExecutionError($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}'.");
+                                }
+                                node.Result = result;
+                            }
+                            catch (Exception error) when (listType.ResolvedType is NonNullGraphType)
+                            {
+                                // Stop execution as early as possible, it doesn't make sense to set the rest of the array elements and then execute nodes
+                                if (error is ExecutionError || defaultException)
+                                    throw;
+                                else
+                                    throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.", error);
+                            }
+                            catch (Exception error) // listType.ResolvedType allows null
```

Comment:
This needs to catch ExecutionError and SetNodeError directly without wrapping, before resorting to this catch block